synapse/handlers/oidc_handler.py

```diff
@@ -977,6 +993,81 @@ def _remote_id_from_userinfo(self, userinfo: UserInfo) -> str:
         return str(remote_user_id)
 
 
+# number of seconds a newly-generated client secret should be valid for
+CLIENT_SECRET_VALIDITY_SECONDS = 3600
+
+# minimum remaining validity on a client secret before we should generate a new one
+CLIENT_SECRET_MIN_VALIDITY_SECONDS = 600
+
+
+class JwtClientSecret:
+    """A class which generates a new client secret on demand, based on a JWK
+
+    This implementation is designed to comply with the requirements for Apple Sign in:
+    https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048
+
+    It looks like those requirements are based on https://tools.ietf.org/html/rfc7523,
+    but it's worth noting that we still put the generated secret in the "client_secret"
+    field (or rather, whereever client_auth_method puts it) rather than in a
+    client_assertion field in the body as that RFC seems to require.
+    """
+
+    def __init__(
+        self,
+        key: OidcProviderClientSecretJwtKey,
+        oauth_client_id: str,
+        oauth_issuer: str,
+        clock: Clock,
+    ):
+        self._key = key
+        self._oauth_client_id = oauth_client_id
+        self._oauth_issuer = oauth_issuer
+        self._clock = clock
+        self._cached_secret = b""
+        self._cached_secret_replacement_time = 0
```

Comment:
I'm not quite sure I understand what would use the cached values here -- does `ClientAuth` cache the secret internally?

Reply:
no it does not. `_get_secret` returns the cached value.

Reply:
I think my question was unclear -- it looked to me like `ClientAuth` was only used once and then discarded, but I see that a single instance is re-used while exchanging the code.

I guess I'm surprised that caching is necessary -- is it very inefficient to calculate the secret each time? Or are you just not supposed to do that?

Reply:
Fair enough! Thanks for the explanation.